src/pgrepr/src/value.rs

```diff
@@ -144,12 +184,26 @@ impl Value {
             Value::Numeric(n) => strconv::format_decimal(buf, &n.0),
             Value::Text(s) => buf.put(s.as_bytes()),
             Value::Jsonb(js) => strconv::format_jsonb(buf, js),
+            Value::List(elems) => {
+                buf.put(&b"{"[..]);
+                let mut elems = elems.iter().peekable();
+                while let Some(elem) = elems.next() {
+                    match elem {
+                        Some(elem) => elem.encode_text(buf),
+                        None => buf.put(&b"NULL"[..]),
+                    }
+                    if elems.peek().is_some() {
+                        buf.put(&b","[..])
+                    }
+                }
+                buf.put(&b"}"[..]);
```

Comment:
Can you put this in `strconv`, so that we'll be able to add a list -> string cast?

Reply:
:+1:

Reply:
Lol, we can't - strconv can't see pgrepr::Value

Reply:
Maybe we can do the conversion from datum to value lazily?

Reply:
Oh, we can't store datum either because of the lifetimes

Reply:
I hate these circular dependency problems. They're not real problems.

Reply:
This one is intentional though! `pgrepr` and `repr` are not supposed to be mutually recursive. `pgrepr` is meant to be a strict layer on top of `repr` that translates from Materialize types to Postgres types and vice-versa.

Reply:
It seems like strconv is relying on the fact that the inner types in pgrepr::Value are the same as the inner types in Datum, but that isn't true for any nested type.

Reply:
Like, format_list should take a DatumList but we already turned it into a Value::List by the time we reach encode. 

Is there a reason encode can't just take a Datum instead of having a separate Value type?

Reply:
Yeah, `strconv` operates on the native Rust types underlying the datums. There are some places where we want to reuse the formatting/parsing logic outside of the dataflow/`Datum` layer, and getting a `Datum` is quite annoying when you have an owned `String` or `Vec<u8>`. I was envisioning something like this:

```diff
diff --git a/src/pgrepr/src/value.rs b/src/pgrepr/src/value.rs
index 3032dccf..b2966c4f 100644
--- a/src/pgrepr/src/value.rs
+++ b/src/pgrepr/src/value.rs
@@ -172,18 +172,7 @@ impl Value {
             Value::Text(s) => buf.put(s.as_bytes()),
             Value::Jsonb(js) => strconv::format_jsonb(buf, js),
             Value::List(elems) => {
-                buf.put(&b"{"[..]);
-                let mut elems = elems.iter().peekable();
-                while let Some(elem) = elems.next() {
-                    match elem {
-                        Some(elem) => elem.encode_text(buf),
-                        None => buf.put(&b"NULL"[..]),
-                    }
-                    if elems.peek().is_some() {
-                        buf.put(&b","[..])
-                    }
-                }
-                buf.put(&b"}"[..]);
+                strconv::format_list(buf, elems, |buf, elem| elem.encode_text(buf))
             }
         }
     }
diff --git a/src/repr/scalar/strconv.rs b/src/repr/scalar/strconv.rs
index b71b430b..a6f0507b 100644
--- a/src/repr/scalar/strconv.rs
+++ b/src/repr/scalar/strconv.rs
@@ -406,6 +406,27 @@ where
     write!(buf, "{:#}", jsonb)
 }
 
+pub fn format_list<'a, F, I, E, L>(buf: &mut F, iter: I, format_element: L)
+where
+    F: FormatBuffer,
+    I: IntoIterator<Item = &'a Option<E>>,
+    E: 'a,
+    L: Fn(&mut F, &'a E),
+{
+    buf.write_char('{');
+    let mut iter = iter.into_iter().peekable();
+    while let Some(elem) = iter.next() {
+        match elem {
+            None => buf.write_str("NULL"),
+            Some(elem) => format_element(buf, elem),
+        }
+        if iter.peek().is_some() {
+            buf.write_char(',');
+        }
+    }
+    buf.write_char('}');
+}
+
 fn format_nanos<F>(buf: &mut F, mut nanos: u32)
 where
     F: FormatBuffer,
```

We'll need to implement that closure for `cast_list_to_string`, but it should just be a quick match on the contained `ScalarType`.

> Is there a reason encode can't just take a Datum instead of having a separate Value type?

A `Value::Decimal` in particular carries with it scale information that a `Datum::Decimal` does not. This is more important on the decode pathway, where pgwire throws more data at us than we can really handle. I personally find it much easier to reason about these things with types that map directly to the data coming over the wire (this crate) and then conversion routines between the two types, than having the mapping from Materialize <-> Postgres intermingled with the pgwire encoding/decoding.